sdk/core/azure-core/azure/core/pipeline/transport/_requests_basic.py

```diff
@@ -130,17 +130,24 @@ def __next__(self):
                 self.response.internal_response.close()
                 raise StopIteration()
             except (requests.exceptions.ChunkedEncodingError,
-                    requests.exceptions.ConnectionError):
+                    requests.exceptions.ConnectionError) as ex:
                 retry_total -= 1
                 if retry_total <= 0:
                     retry_active = False
                 else:
                     time.sleep(retry_interval)
                     headers = {'range': 'bytes=' + str(self.downloaded) + '-'}
-                    resp = self.pipeline.run(self.request, stream=True, headers=headers)
-                    if resp.http_response.status_code == 416:
-                        raise
-                    chunk = next(self.iter_content_func)
+                    try:
+                        resp = self.pipeline.run(self.request, stream=True, headers=headers)
+                        if not resp.http_response:
+                            raise
+                        if resp.http_response.status_code == 416:
+                            raise
+                        chunk = next(self.iter_content_func)
```

Comment:
There is no point calling `next(self.iter_content_func)` at this point, as this line is only reached if the previous call to `next(self.iter_content_func)` in L124 raised an exception.

Note that this follows from the specificaiton of python generators. `self.iter_content_func` is a python generator and according to the python generator specification [1]:

```
If an unhandled exception [...] is raised by, or passes through, a generator function, then the exception is passed on to the caller in the usual way, and subsequent attempts to resume the generator function raise StopIteration. In other words, an unhandled exception terminates a generator's useful life.
```

Therefore: please remove this call. It has no purpose.


[1] https://www.python.org/dev/peps/pep-0255/#specification-generators-and-exception-propagation

Reply:
@jochen-ott-by Thanks for the comments.

The logic behind it is:

If there is a ConnectionError raises during the downloading, we will wait for a while,
```py
time.sleep(retry_interval)
```

Then try to re-connect and continue the downloading by setting the header
```py
headers = {'range': 'bytes=' + str(self.downloaded) + '-'}
```

and yes, if we hit failures again, we will pop it up to the caller (this is what we fix in this PR).

Reply:
Thanks for the answer. However, your description does not match with what the code does.

"Then try to re-connect and continue the downloading" this is misleading, as it's *not* a new connection that is used for the attempt to continue the downloading. Rather, the original value of `self.iter_content_func` is used, the one that just raised an exception and can no longer be used for anything that makes sense.
The only failure reported here is a 416 "range not satisfiable".

And as already commented, the code makes assumptions it cannot safely make: StreamDownloader is also used for requests that already have a (non-trivial) range header set. So what happens if the original range header is, say, "bytes=1000-1999", chunking is 200 and there is a connection error after the first chunk?
This code will then go on a make a request for "range: bytes=200-", which does not overlap with the original request.

This makes no sense either.